[JIT] Add mutability checks in FunctionSchema and create SchemaInfo subclass

ttanpcs opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81007
* #81002
* #81000
* #80984
* #80972
* #80918
* __->__ #80734

- Added overloads to is_mutable method in FunctionSchema to tell whether an argument at index is mutable or an argument with name is mutable.
- Created SchemaInfo subclass of FunctionSchema with constructors from FunctionSchema and from const char* signature.
- Tested is_mutable method overloads in new test_schema_info.cpp file. 

**Note that this pr is used to set up SchemaInfo. Implementation for SchemaInfo will be addressed in later commits**

Differential Revision: [D37651384](https://our.internmc.facebook.com/intern/diff/D37651384)

Commit:
[JIT] Add mutability checks in FunctionSchema and create SchemaInfo subclass

[ghstack-poisoned]

Commit:
Update on "[JIT] Add mutability checks in FunctionSchema and create SchemaInfo subclass"




[ghstack-poisoned]

Commit:
Update on "[JIT] Add mutability checks in FunctionSchema and create SchemaInfo subclass"




[ghstack-poisoned]

goldenxuett commented:
@goldenxuett has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37651384).

Commit:
Update on "[JIT] Add mutability checks in FunctionSchema and create SchemaInfo subclass"


- Added overloads to is_mutable method in FunctionSchema to tell whether an argument at index is mutable or an argument with name is mutable.
- Created SchemaInfo subclass of FunctionSchema with constructors from FunctionSchema and from const char* signature.
- Tested is_mutable method overloads in new test_schema_info.cpp file. 

**Note that this pr is used to set up SchemaInfo. Implementation for SchemaInfo will be addressed in later commits**

Differential Revision: [D37651384](https://our.internmc.facebook.com/intern/diff/D37651384)

[ghstack-poisoned]

goldenxuett commented:
@goldenxuett has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37651384).

davidberard98 reviewed (approved):
LGTM!

Commit:
Update on "[JIT] Add mutability checks in FunctionSchema and create SchemaInfo subclass"


- Added overloads to is_mutable method in FunctionSchema to tell whether an argument at index is mutable or an argument with name is mutable.
- Created SchemaInfo subclass of FunctionSchema with constructors from FunctionSchema and from const char* signature.
- Tested is_mutable method overloads in new test_schema_info.cpp file. 

**Note that this pr is used to set up SchemaInfo. Implementation for SchemaInfo will be addressed in later commits**

Differential Revision: [D37651384](https://our.internmc.facebook.com/intern/diff/D37651384)

[ghstack-poisoned]

Commit:
Update on "[JIT] Add mutability checks in FunctionSchema and create SchemaInfo subclass"


- Added overloads to is_mutable method in FunctionSchema to tell whether an argument at index is mutable or an argument with name is mutable.
- Created SchemaInfo subclass of FunctionSchema with constructors from FunctionSchema and from const char* signature.
- Tested is_mutable method overloads in new test_schema_info.cpp file. 

**Note that this pr is used to set up SchemaInfo. Implementation for SchemaInfo will be addressed in later commits**

Differential Revision: [D37651384](https://our.internmc.facebook.com/intern/diff/D37651384)

[ghstack-poisoned]

goldenxuett commented:
@goldenxuett has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37651384).

davidberard98 commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -359,7 +359,20 @@ struct FunctionSchema {
           return aliasInfo && aliasInfo->isWrite();
         });
   }
+  bool is_mutable(size_t index) const {
```
actually can you make these camel case like the other methods in this class?

Gamrix commented on torch/csrc/utils/schema_info.h:
```diff
@@ -0,0 +1,24 @@
+/**
+ * class SchemaInfo
+ *
+ * Subclass of FunctionSchema that publicizes argument value specific operator
+ * behavior (mutation, aliasing, special cases, etc...)
+ */
+
+struct TORCH_API SchemaInfo : c10::FunctionSchema {
+ public:
+  explicit SchemaInfo(c10::FunctionSchema schema) : FunctionSchema(schema) {}
+  explicit SchemaInfo(const char* signature)
+      : FunctionSchema(torch::jit::getOperatorForLiteral(signature)->schema()) {
+  }
+};
```
@davidberard98 Should we be concerned about performance impacts from virtualizing FunctionSchema? FunctionSchema is used in Core.

bdhirsh replied:
Are any methods on `FunctionSchema` actually getting virtualized in this PR? It looks like `SchemaInfo inherits from `FunctionSchema` (but without virtualizing any methods). I think that should be fine.

Gamrix replied:
@bdhirsh Look at https://github.com/pytorch/pytorch/pull/80972/files for functions in `SchemaInfo` that we would be virtualizing. (They are currently not marked as virtual, but I expect that we would need them to be `virtual` in order for things to work as intended).

bdhirsh replied:
Ah ok, in agreement with John then on composition over inheritance. `FunctionSchema` is probably hot-path enough that we want to avoid virtualizing some of its common methods.

Commit:
Update on "[JIT] Add mutability checks in FunctionSchema and create SchemaInfo subclass"


- Added overloads to is_mutable method in FunctionSchema to tell whether an argument at index is mutable or an argument with name is mutable.
- Created SchemaInfo subclass of FunctionSchema with constructors from FunctionSchema and from const char* signature.
- Tested is_mutable method overloads in new test_schema_info.cpp file. 

**Note that this pr is used to set up SchemaInfo. Implementation for SchemaInfo will be addressed in later commits**

Differential Revision: [D37651384](https://our.internmc.facebook.com/intern/diff/D37651384)

[ghstack-poisoned]

goldenxuett commented:
@goldenxuett has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37651384).

Commit:
Update on "[JIT] Add mutability checks in FunctionSchema and create SchemaInfo subclass"


- Added overloads to is_mutable method in FunctionSchema to tell whether an argument at index is mutable or an argument with name is mutable.
- Created SchemaInfo subclass of FunctionSchema with constructors from FunctionSchema and from const char* signature.
- Tested is_mutable method overloads in new test_schema_info.cpp file. 

**Note that this pr is used to set up SchemaInfo. Implementation for SchemaInfo will be addressed in later commits**

Differential Revision: [D37651384](https://our.internmc.facebook.com/intern/diff/D37651384)

[ghstack-poisoned]

goldenxuett commented:
@goldenxuett has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37651384).

davidberard98 reviewed (approved):
looks good!
